root-drive-science-ios/ViewController.swift

```diff
@@ -102,13 +125,42 @@ class ViewController: UIViewController {
 
     private func checkLocationServicesEnabled() {
         if !CLLocationManager.locationServicesEnabled() {
-            appendNotificationText("Location services disabled at system level! Will be unable to track trips!")
+            addLogMessage("Location services disabled at system level! Will be unable to track trips!")
+        }
+    }
+
+    func addLogMessage(_ text: String) {
+        if let newMessage = saveLogMessage(text) {
+            displayLogMessage(newMessage)
         }
     }
 
-    func appendNotificationText(_ text: String) {
+    private func saveLogMessage(_ text: String) -> LogMessage? {
+        let managedContext = container.viewContext
+
+        let entity = NSEntityDescription.entity(forEntityName: "LogMessage", in: managedContext)!
+
+        let logMessage = LogMessage(entity: entity, insertInto: managedContext)
+
+        logMessage.message = text
+        logMessage.date = Date()
+
+        do {
+            try managedContext.save()
+            logMessages.append(logMessage)
+            return logMessage
+        } catch let error as NSError {
+            print("Could not save log message. \(error), \(error.userInfo)")
+        }
+
+        return nil
+    }
+
+    private func displayLogMessage(_ logMessage: LogMessage) {
         DispatchQueue.main.async {
-            self.notificationField.text = self.notificationField.text.appending("[\(self.timestamp)]:" + text + "\n")
+            let message = logMessage.message!
```

Comment:
The force unwrap here is dangerous and will cause crashes if logMessage ever has a nil message or date. Would recommend wrapping this in an if/let

```
if let message = logMessage.message, let date = logMessage.date {
    self.notificationField.text = // ...
}
```

It is also possible to make these fields required in the data model metadata and then the properties would not be optional.

Reply:
👍